src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Numerics;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    /// <summary>
+    /// Middleware that logs HTTP requests and HTTP responses.
+    /// </summary>
+    internal class W3CLoggingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly W3CLogger _w3cLogger;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+        private string? _serverName;
+        private bool _hasLogged;
+
+        internal static Dictionary<W3CLoggingFields, int> _fieldIndices = new Dictionary<W3CLoggingFields, int>()
+        {
+            { W3CLoggingFields.Date, BitOperations.Log2((int)W3CLoggingFields.Date) },
+            { W3CLoggingFields.Time, BitOperations.Log2((int)W3CLoggingFields.Time) },
+            { W3CLoggingFields.ClientIpAddress, BitOperations.Log2((int)W3CLoggingFields.ClientIpAddress) },
+            { W3CLoggingFields.UserName, BitOperations.Log2((int)W3CLoggingFields.UserName) },
+            { W3CLoggingFields.ServerName, BitOperations.Log2((int)W3CLoggingFields.ServerName) },
+            { W3CLoggingFields.ServerIpAddress, BitOperations.Log2((int)W3CLoggingFields.ServerIpAddress) },
+            { W3CLoggingFields.ServerPort, BitOperations.Log2((int)W3CLoggingFields.ServerPort) },
+            { W3CLoggingFields.Method, BitOperations.Log2((int)W3CLoggingFields.Method) },
+            { W3CLoggingFields.UriStem, BitOperations.Log2((int)W3CLoggingFields.UriStem) },
+            { W3CLoggingFields.UriQuery, BitOperations.Log2((int)W3CLoggingFields.UriQuery) },
+            { W3CLoggingFields.ProtocolStatus, BitOperations.Log2((int)W3CLoggingFields.ProtocolStatus) },
+            { W3CLoggingFields.TimeTaken, BitOperations.Log2((int)W3CLoggingFields.TimeTaken) },
+            { W3CLoggingFields.ProtocolVersion, BitOperations.Log2((int)W3CLoggingFields.ProtocolVersion) },
+            { W3CLoggingFields.Host, BitOperations.Log2((int)W3CLoggingFields.Host) },
+            { W3CLoggingFields.UserAgent, BitOperations.Log2((int)W3CLoggingFields.UserAgent) },
+            { W3CLoggingFields.Cookie, BitOperations.Log2((int)W3CLoggingFields.Cookie) },
+            { W3CLoggingFields.Referer, BitOperations.Log2((int)W3CLoggingFields.Referer) }
```

Comment:
Can these be `const`s or `private static readonly int`s?

Reply:
What's the upside? I like that using the collection allows me to set `_fieldsLength = _fieldIndices.Count`

Reply:
Avoiding an extra dictionary lookup for every key for every request when it could be a constant. It's not like it's a huge cost, but it's feels completely unnecessary.

Reply:
Fair enough, pushed an update